### PR TITLE
Begin Receiving and Processing Blocks for Optimized Slasher

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR begins adding the methods to receive and process batches of beacon block headers for slashing detection as part of our optimized slasher. Double proposal detection is somewhat trivial and can be completed in the span of a few pull requests. This pull requests implements the receiver functionality from an event feed of beacon block headers similar to how we handle attestations as well as simple unit tests. 

The actual detection logic can be done in a separate pull request.

**Which issues(s) does this PR fix?**

Part of #8331

### CODE REVIEW

File: beacon-chain/slasher/receive.go [hunk 1]
```diff
@@ -27,9 +27,34 @@ func (s *Service) receiveAttestations(ctx context.Context) {
...
+
+// Receive beacon blocks from some source event feed,
+func (s *Service) receiveBlocks(ctx context.Context) {
+	sub := s.serviceCfg.BeaconBlocksFeed.Subscribe(s.beaconBlocksChan)
+	defer close(s.beaconBlocksChan)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case blockHeader := <-s.beaconBlocksChan:
+			// TODO(#8331): Validate block and defer blocks from the future for later processing.
```
Comment (terencechain) [inline beacon-chain/slasher/receive.go:50]: We dont need to validate the block for it to become slashable. A proposer can sign duplicated bad blocks and still be slashed

File: beacon-chain/slasher/receive_test.go [hunk 2]
```diff
@@ -5,8 +5,12 @@ import (
 	"testing"
 
 	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	logTest "github.com/sirupsen/logrus/hooks/test"
+
```
Comment (terencechain) [inline beacon-chain/slasher/receive_test.go:9]: line lol

File: beacon-chain/slasher/types/types.go [hunk 3]
```diff
@@ -15,11 +15,19 @@ const (
...
+	Target           types.Epoch
 	SigningRoot      [32]byte
 }
 
+// CompactBeaconBlock containing only the required information
+// for proposer slashing detection.
+type CompactBeaconBlock struct {
+	ProposerIndex uint64
+	Slot          uint64
+	SigningRoot   [32]byte
```
Comment (terencechain) [inline beacon-chain/slasher/types/types.go:28]: Why do we need `SigningRoot`? 

If it's needed shouldn't we have multiple `SigningRoot`?

File: beacon-chain/slasher/receive.go [hunk 4]
```diff
@@ -78,3 +105,23 @@ func (s *Service) processQueuedAttestations(ctx context.Context, epochTicker <-c
 		}
 	}
 }
+
+// Process queued blocks every time an epoch ticker fires. We retrieve
```
Comment (terencechain) [inline beacon-chain/slasher/receive.go:109]: Keep in mind this has liveness concern to always wait one epoch. One can say that prysm is less performant at broadcasting or holding/proposing proposer slashing.

Major exchanges/whales may not like this

File: beacon-chain/slasher/receive.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/slasher/receive.go:50]: We just need to ensure that this passes

https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#proposer-slashings

File: beacon-chain/slasher/receive.go (hunk 4 above)
Comment (rauljordan) [reply beacon-chain/slasher/receive.go:109]: Indeed, per slot seems like is best but we want to better benchmark slasher's performance first. We will explore changing these values in future PRs that look at performance

File: beacon-chain/slasher/types/types.go (hunk 3 above)
Comment (rauljordan) [reply beacon-chain/slasher/types/types.go:28]: this is a compact beacon block, not a proposer slashing. We need signing root to determine if a block is slashable with respect to another one, namely that the other fields match except for the signing root

**Review by terencechain — APPROVED** [review]

**Pull request merged**